Support returning `!u8` from main

SamTebbs33 opened the pull request:
This patch assigns the return value of `root.main()` to a variable to allow returning `!u8` from a program's main function.

Fixes #2721 and #2662

tiehuis commented on std/special/bootstrap.zig:
```diff
@@ -123,12 +123,15 @@ inline fn callMain() u8 {
         builtin.TypeId.ErrorUnion => {
-            root.main() catch |err| {
+            const PayloadType = @typeOf(root.main).ReturnType.Payload;
+            if (@typeId(PayloadType) != builtin.TypeId.Void and @typeId(PayloadType) != builtin.TypeId.Int)
```
This `Int` check is a bit lenient. I would check if it is exactly a u8 explicitly. This will be caught on the return of this function still, but you would not get the nicer compile error.

This actually occurs right now if you use an `i8` as the return value for the above branch.

SamTebbs33 replied:
Thanks for spotting that

tiehuis commented:
You'll need to update the compile error test cases as well. See here:
https://github.com/ziglang/zig/blob/2cbcf3f38bb32752b449ce72a8d8f978e65008f8/test/compile_errors.zig#L2187

SamTebbs33 commented:
> You'll need to update the compile error test cases as well. See here:
> https://github.com/ziglang/zig/blob/2cbcf3f38bb32752b449ce72a8d8f978e65008f8/test/compile_errors.zig#L2187

I should probably add a test to make sure a main function returning `!u8` is always allowed, what file would be best for that?

andrewrk commented:
> I should probably add a test to make sure a main function returning `!u8` is always allowed, what file would be best for that?

Probably a "standalone" test. Add an entry here: https://github.com/ziglang/zig/blob/master/test/build_examples.zig

SamTebbs33 commented:
> > I should probably add a test to make sure a main function returning `!u8` is always allowed, what file would be best for that?
> 
> Probably a "standalone" test. Add an entry here: https://github.com/ziglang/zig/blob/master/test/build_examples.zig

Added a standalone test and updated some compile error tests to look for the new error message.

marler8997 commented:
I had made another PR for this not knowing you had made one.  I have 3 suggetsions:

1. in the ErrorUnion and Int case, you should probably also check that the integer is 8 bits
2. the error message is in 3 to 4 places now, might be better to assign it to a variable
3. you may also want to add the actual return type to the error message

These 3 suggestions were implemented in my PR.  You can reference my branch if you'd like to see how I implemented them:

My Branch: https://github.com/marler8997/zig/commit/f0ae570bfe84eb853c268539d68603691f84da95
My Closed PR: https://github.com/ziglang/zig/pull/2800

marler8997 commented on std/special/bootstrap.zig:
```diff
@@ -139,7 +142,7 @@ inline fn callMain() u8 {
             };
             return 0;
```
This is a bug.  if main returned a non-zero value, you're ignoring it.

SamTebbs33 replied:
Thanks for noticing that! Fixed.

marler8997 commented on test/standalone/main_return_error/!u8.zig:
```diff
@@ -0,0 +1,7 @@
+const Err = error {
```
You should also add a test case where main returns a non-zero value.

SamTebbs33 replied:
I have added one but it needs something to check that its exit code is correct, do you know how I would do so?

marler8997 replied:
Sorry I'm not familiar with the test framework.  I wish I was more familiar.

marler8997 commented:
Also you'll need to rebase since `bootstrap.zig` got renamed to `start.zig`.

marler8997 commented on std/special/bootstrap.zig:
```diff
@@ -123,23 +125,30 @@ inline fn callMain() u8 {
-                        std.debug.dumpStackTrace(trace.*);
+            const PayloadType = @typeOf(root.main).ReturnType.Payload;
+            // In this case the error should include the payload type
+            const payload_err = "expected return type of main to be 'u8', 'noreturn', 'void', '!void', or '!u8', found '!" ++ @typeName(PayloadType) ++ "'";
```
why not use `compile_err`?

SamTebbs33 replied:
Because `compile_err` uses the name of the return type, which is fine when it isn't an error union since then it looks like `foo.bar.ErrorUnion!T` so I use a different variable when it is an error union. I could make `compile_err` a `var` and reassign it.

marler8997 replied:
I just don't like the duplication, how about:
```zig
const compile_err_prefix = "expected return type of main to be 'u8', 'noreturn', 'void', '!void', or '!u8'"
...
// in the payload case
const compile_err = compile_err_prefix ++ ", found '!" ++ @typeName(PayloadType) ++ "'";
...
// in the other cases
const compile_err = compile_err_prefix ++ ", found '" ++ @typeName(@typeOf(root.main).ReturnType) ++ "'";
```

marler8997 commented on std/special/bootstrap.zig:
```diff
@@ -123,23 +125,30 @@ inline fn callMain() u8 {
+            const payload_err = "expected return type of main to be 'u8', 'noreturn', 'void', '!void', or '!u8', found '!" ++ @typeName(PayloadType) ++ "'";
+            // If the payload is void or a u8
+            if (@typeId(PayloadType) == builtin.TypeId.Void or (@typeId(PayloadType) == builtin.TypeId.Int and PayloadType.bit_count == 8)) {
+                const tmp: PayloadType = root.main() catch |err| {
```
*nit, no need for `: PayloadType`

marler8997 commented on test/build_examples.zig:
```diff
@@ -7,6 +7,7 @@ pub fn addCases(cases: *tests.BuildExamplesContext) void {
     cases.addC("example/hello_world/hello_libc.zig");
     cases.add("example/cat/main.zig");
     cases.add("example/guess_number/main.zig");
+    cases.add("test/standalone/main_return_error/!u8.zig");
```
Having `!` in the filename may cause issues on some filesystems.

SamTebbs33 commented:
> thanks for implementing all my suggestions

Thanks for the suggestions! They made the PR a lot better
